compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1673,92 +1677,101 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         };
 
         debug!("one_bound_for_assoc_type: bound = {:?}", bound);
-
-        if let Some(bound2) = matching_candidates.next() {
-            debug!("one_bound_for_assoc_type: bound2 = {:?}", bound2);
-
-            let is_equality = is_equality();
-            let bounds = IntoIterator::into_iter([bound, bound2]).chain(matching_candidates);
-            let mut err = if is_equality.is_some() {
-                // More specific Error Index entry.
-                struct_span_err!(
-                    self.tcx().sess,
-                    span,
-                    E0222,
-                    "ambiguous associated type `{}` in bounds of `{}`",
-                    assoc_name,
-                    ty_param_name()
-                )
-            } else {
-                struct_span_err!(
-                    self.tcx().sess,
-                    span,
-                    E0221,
-                    "ambiguous associated type `{}` in bounds of `{}`",
-                    assoc_name,
-                    ty_param_name()
-                )
-            };
-            err.span_label(span, format!("ambiguous associated type `{}`", assoc_name));
-
-            let mut where_bounds = vec![];
-            for bound in bounds {
-                let bound_id = bound.def_id();
-                let bound_span = self
-                    .tcx()
-                    .associated_items(bound_id)
-                    .find_by_name_and_kind(self.tcx(), assoc_name, ty::AssocKind::Type, bound_id)
-                    .and_then(|item| self.tcx().hir().span_if_local(item.def_id));
-
-                if let Some(bound_span) = bound_span {
-                    err.span_label(
-                        bound_span,
-                        format!(
-                            "ambiguous `{}` from `{}`",
+        let next_cand = if is_ty { matching_candidates.next() } else { const_candidates.next() };
```

Comment:
instead of having this `is_ty` variable, return `next_cand` directly from the match where you return `bound`.